docs/content/v0.20/architecture.md

```diff
@@ -1,36 +1,479 @@
+    cache (which is kept updated by the reflector) by the time a worker
+    gets a copy from the work queue. Workers should always receive the
+    key and use it to retrieve the object from the cache.
+
+3.  A worker pulls a key from the work queue, and then does the
+    following processing:
+
+    -  Uses the GVK key to get the reference to the lister for the
+        object
+    -  Gets the object from the lister cache using the key that was
+        generated with the MetaNamespaceKeyFunc.
+    -  Gets the lister for Placement objects and list all placements
+    -  Iterates on all placements, and for each placement:
+        - evaluates the placement label selector vs. object labels to
+          check if there is a match.
+        - If there is a match, uses the lister for ManagedCluster and
+          find the matching clusters using the label selector expression
+          for clusters.
+        - If there are matching clusters, add the names of the cluster
+          to a hashmap setting the name of the cluster as a key. Cluster
+          groups from different placements are merged together.
+    -  If there are no matching clusters, the worker returns without
+        actions and is ready to process other events from the queue.
+    -  If there are matching clusters:
+        - Wraps the object into a ManifestWork
+        - Adds a label to the ManifestWork that is used to track the
+          placement that caused the object to be delivered to one or
+          more clusters. The label contains both the placement name
+          (note that placement is cluster-scoped) and the WDS name. A
+          label is attached for each placement whose selector matches
+          the object. This way, when a placement is deleted or
+          updated it is possible to locate the associated ManifestWorks
+          for deletion or label removal (if more than one label is
+          present, as other placements may “own” the object).
+        - For each cluster:
+          -  Sets the manifest namespace == name of the cluster
+          -  Uses the client connected to the ITS to do a server-side
+              apply of the manifest into the namespace.
+        - Worker returns and is ready to pick other keys from the queue.
+
+There are other event flows, based on the object GVK and type of event.
+The following sections broadly describe these flows:
+
+#### Object Deleted 
+
+When an object is deleted from the WDS, the handler’s *DeleteFunc* is
+invoked. Since most objects do not have a finalizer (which prevents
+deletion from the API server until the controller that sets the
+finalizer does some cleanup and then removes the object), a shallow copy
+of the object is added to a field in the key before pushing it to the
+work queue. Then:
+
+- Worker pulls the key from the work queue
+
+- Flow continues the same way as in the create/update scenario, however
+  the deletedObject field in the key indicates that the object has been
```

Comment:
What if, at this time, there is now an object matching the key pulled from the work queue?

Reply:
If the key has a not null `DeletedObject` it will not get the object from the cache (thus it will not get the new object at this time), therefore it will go to the delete flow and that object gets deleted. The creation of a new object with same (GVK,namespace,name) right after the deletion of an object with the same (GVK,namespace,name) will trigger a create event and event handler will create a new key with a null `DeletedObject` which would then be handled like a new object to evaluate vs. placements.

Reply:
Yes, this is functionally correct but has a performance down-side of doing unnecessary work.